### ISSUE TITLE
org.clojure/clojurescript at 38.7% in report for react-native project

### ISSUE THREAD

**Issue by NuclearKev** [thheller/shadow-cljs#523]
Hi guys, I'm fairly new to using shadow-cljs and was a little concerned about the size of the optimized `org.clojure/clojurescript` bundle. 

I'm currently using shadow-cljs for a react native project and am running into some performance issues with startup on android. So, I looked into ways to make startup quicker, most notably, [this](https://facebook.github.io/react-native/docs/performance) article. After doing the RAM bundle and inline requires stuff, I saw a big difference but not big enough. So I looked into preloading modules. In that article, they mention a way to print out the modules so you can preload them. The number of loaded modules was huge and there were tons of CLJS libs listed as well. At this point I tried running a build report which backed up my findings: CLJS is loading a ton of modules at startup. 

To test this farther, I made a RN project without CLJS and a project with CLJS - the project just being a white page with some text. The RN project started up in about 1s whereas the CLJS one took about 3s. This makes me assume that it's the loading of CLJS libs that is making startup so long. 

Does anyone know if there is a way to minimize the size of the CLJS libs (more than advanced optimizations) or remove code that isn't needed inside those libs (I know about code stripping but that seems dangerous)? Thanks!

**Comment by thheller** [comment]
CLJS itself will always be a certain size given that `cljs.core` basically contains the equivalent of several JS libs (eg. `immutable-js`, `lodash`, etc). What isn't used will be removed by `:advanced`  so it should be comparable. It obviously is not comparable if you compare to plain JS without any other dependencies. Actuall app startup time shouldn't be that much different for real apps.

There are certain optimizations that can certainly be done using the regular ClojureScript code-splitting (`:modules`) but I need to spend time time with react-native first to figure out how it all works on their side. Seems like it is a rather manual and hacky process currently.

**Comment by NuclearKev** [comment]
Hi @thheller, we are still going strong with our cljs RN app using shadow-cljs! However, I've never quite gotten this part figured out. The load time is bearable and really wasn't much of a concern until now. We are planning on doing some whitelabelling for our app and would like there to be different "splash screens" for the customers who enable this feature. I put splash screen in quotes because we aren't actually changing the app's splash screen but rather the loading screen after it. Which brings me to my question: is it possible to use code-splitting with `:modules` on react native? I've tried using it as  your article describes but shadow fails to find the module that I test it with. Any ideas?

**Comment by thheller** [comment]
No clue. I have not looked at react-native in forever. I'm fairly certain that it is possible, just don't ask me how. If you have a JS example I can reference I can make suggestions.

**Comment by NuclearKev** [comment]
Here is the shadow-cljs.edn: 
```
{:deps {:aliases [:dev]}
 :nrepl {:port 9000}

 :builds
 {:app
  {:target :react-native
   :init-fn sked-app.core/init
   :output-dir "app"
   :compiler-options {:infer-externs :auto}
   :release {:compiler-options {:strip-type-prefixes #{"cljs.pprint"
                                                       "cljs.spec.test.alpha"}}}
   :dev {:compiler-options {:closure-defines {"re_frame.trace.trace_enabled_QMARK_" true}}}
   :devtools {:autoload true}
   :module-loader true
   :modules
   {:main
    {:init-fn sked-app.core/init}

    :login
    {:entries [sked-app.pages.login.view]
     :depends-on #{:main}}}}}}
```

Then I get this error:
```
[:app] Build failure:
------ ERROR -------------------------------------------------------------------
 File: /home/kev/work/app/src/sked_app/core.cljs:88:12
--------------------------------------------------------------------------------
  85 |             [sked-app.components.progress :refer [progress-bar]]
  86 |             [shadow.lazy :as lazy]))
  87 | 
  88 | (def login (lazy-component sked-app.pages.login.view/login))
------------------^-------------------------------------------------------------
Encountered error when macroexpanding shadow.lazy/loadable.
Could not find module for ns: sked-app.pages.login.view at line 88 sked_app/core.cljs
--------------------------------------------------------------------------------
  89 | (def sign-in (lazy-component sked-app.pages.login.view/sign-in))
  90 | (def login-email (lazy-component sked-app.pages.login.view/login-email))
  91 | (def forgot-password (lazy-component sked-app.pages.login.view/forgot-password))
  92 | (def password-reset-sent (lazy-component sked-app.pages.login.view/password-reset-sent))
--------------------------------------------------------------------------------
```
`sked-app.pages.login.view` is correct and, as you can see, is in the modules. Any ideas why I might be getting this?

**Comment by thheller** [comment]
The `shadow.lazy` helpers will not work as they use browser APIs. This is also the part I cannot answer how you'd use it in RN, since I do not know the JS side of things.

What I can say that code splitting works differently at the config level already. `:modules` doesn't work. Instead you have `:chunks`, which is essentially `:modules` but with an implicit depency on the main "index" module.

Adjusting your config that looks something like this:

```
{:deps {:aliases [:dev]}
 :nrepl {:port 9000}

 :builds
 {:app
  {:target :react-native
   :init-fn sked-app.core/init
   :output-dir "app"
   :dev {:compiler-options {:closure-defines {re-frame.trace/trace-enabled? true}}}
   :chunks {:login sked-app.pages.login.view/x}}}}
```

Basically you get a `index.js` in `:output-dir` as usual, plus an additional `login.js` that when loaded will yield whatever the `sked-app.pages.login.view/x` var holds. How you actually trigger the load I have no idea. It used to be `(js/require "./login.js")` I believe. I never actually tried any of this in an actual RN app.

**Comment by NuclearKev** [comment]
> The `shadow.lazy` helpers will not work as they use browser APIs. This is also the part I cannot answer how you'd use it in RN, since I do not know the JS side of things.
> 
> What I can say that code splitting works differently at the config level already. `:modules` doesn't work. Instead you have `:chunks`, which is essentially `:modules` but with an implicit depency on the main "index" module.
> 
> Adjusting your config that looks something like this:
> 
> ```
> {:deps {:aliases [:dev]}
>  :nrepl {:port 9000}
> 
>  :builds
>  {:app
>   {:target :react-native
>    :init-fn sked-app.core/init
>    :output-dir "app"
>    :dev {:compiler-options {:closure-defines {re-frame.trace/trace-enabled? true}}}
>    :chunks {:login sked-app.pages.login.view/x}}}}
> ```
> 
> Basically you get a `index.js` in `:output-dir` as usual, plus an additional `login.js` that when loaded will yield whatever the `sked-app.pages.login.view/x` var holds. How you actually trigger the load I have no idea. It used to be `(js/require "./login.js")` I believe. I never actually tried any of this in an actual RN app.

This works! However, even if I did this splitting for everything (I tested it with the larger files) I would likely only drop my bundle to maybe 1.1MB which might help the speed a little bit. Seems like maybe the squeeze really isn't worth the juice at least on the RN side of things. Not to mention, there seems to be issues when building the apps for production (being unable to find the chunked JS files). 

Thanks for helping me out!

**Comment by thheller** [comment]
I guess the minimal variant would be writing the splash screen in JS and have that lazy load the rest. That would at least push the cljs.core parts into the lazy bit, which is likely a big chunk overall.
